engine/class_modules/paladin/sc_paladin_protection.cpp

```diff
@@ -67,36 +55,98 @@ struct avengers_shield_t : public paladin_spell_t
+    {
+      p() -> buffs.soaring_shield -> trigger();
+    }
 
-    base_multiplier *= 1.0 + p -> talents.first_avenger -> effectN( 1 ).percent();
-    // First Avenger only increases primary target damage
-    base_aoe_multiplier *= 1.0 / ( 1.0 + p -> talents.first_avenger -> effectN( 1 ).percent() );
+    // First Avenger absorb shield. Amount is additive per hit.
+    if ( p() -> talents.first_avenger -> ok() )
+    {
+      //First Avenger is capped at 30% of max hp. Hardcoded here.
+      double max_absorb = 0.3 * p() -> resources.max[ RESOURCE_HEALTH ];
 
-    // link needed for trigger_grand_crusader
+      double new_absorb = s -> result_amount * p() -> talents.first_avenger -> effectN( 2 ).percent();
+      if( p() -> buffs.first_avenger_absorb -> value() + new_absorb < max_absorb )
+        p() -> buffs.first_avenger_absorb -> trigger( 1, p() -> buffs.first_avenger_absorb -> value() + new_absorb );
+      else
+        p() -> buffs.first_avenger_absorb -> trigger( 1, max_absorb );
+    }
+
+    if ( p() -> conduit.vengeful_shock -> ok() )
+    {
+      td( s -> target ) -> debuff.vengeful_shock -> trigger();
+    }
+  }
+};
+
+struct avengers_shield_dt_t : public avengers_shield_base_t
+{
+  avengers_shield_dt_t( paladin_t* p ) :
+    avengers_shield_base_t( "avengers_shield_dt", p, p -> find_specialization_spell( "Avenger's Shield" ), "" )
+  {
+    background=true;
+  }
+};
+
+struct avengers_shield_t : public avengers_shield_base_t
+{
+  avengers_shield_t( paladin_t* p, const std::string& options_str ) :
+    avengers_shield_base_t( "avengers_shield", p, p -> find_specialization_spell( "Avenger's Shield" ), options_str )
+  {
     cooldown = p -> cooldowns.avengers_shield;
   }
 
   void execute() override
   {
-    paladin_spell_t::execute();
-
-    p() -> buffs.avengers_valor -> trigger();
-    if ( p() -> talents.redoubt -> ok() )
+    avengers_shield_base_t::execute();
+    if( p() -> buffs.moment_of_glory -> up() )
     {
-      p() -> buffs.redoubt -> trigger();
+      p() -> cooldowns.avengers_shield -> reset( false );
+      p() -> buffs.moment_of_glory -> decrement( 1 );
     }
   }
 
-  void impact( action_state_t* s ) override
+  double action_multiplier() const override
   {
-    paladin_spell_t::impact( s );
+    double m = avengers_shield_base_t::action_multiplier();
+    if( p() -> buffs.moment_of_glory -> up() )
+      m *= 1.0 + p() -> buffs.moment_of_glory -> value();
 
-    if ( p() -> azerite.soaring_shield.enabled() )
-    {
-      p() -> buffs.soaring_shield -> trigger();
-    }
+    return m;
+  }
+};
+
+// Moment of Glory ============================================================
+
+struct moment_of_glory_t : public paladin_spell_t
+{
+  moment_of_glory_t( paladin_t* p, const std::string& options_str ) :
+    paladin_spell_t( "moment_of_glory", p, p -> talents.moment_of_glory )
+  {
+    parse_options( options_str );
+
+    harmful = false;
+    use_off_gcd = true;
```

Comment:
Same as for Avenging Wrath. 
trigger_gcd shoulldn't need to be set to 0_ms if the correct spelldata is used.